gocontext.go

```diff
@@ -47,7 +47,10 @@ func StartSpan(ctx context.Context, name, spanType string) (*Span, context.Conte
 // when the span completes.
 func StartSpanOptions(ctx context.Context, name, spanType string, opts SpanOptions) (*Span, context.Context) {
 	tx := TransactionFromContext(ctx)
-	span := tx.StartSpan(name, spanType, SpanFromContext(ctx))
+	if opts.Parent == (TraceContext{}) {
```

Comment:
I'm rather embarrassed about this bug, thanks for picking it up!

There's a couple of issues here. As it is, this code will panic if ctx does not contain a transaction. This can be observed in the test failures. Second, StartSpanOptions is supposed to use the parent span from ctx, if it contains one. It would have done that before, but with this change it won't.

I think how it should be is this:
 - If opts.Parent is non-zero, leave opts alone
 - Otherwise, set opts.Parent to the trace context of the span in ctx, if any

This should work:

```go
if opts.Parent == (TraceContext{}) {
    opts.Parent = SpanFromContext(ctx).TraceContext()
}
```

(*Span.TraceContext() will return a zero trace context value if the span is nil.)

The call to tx.StartSpanOptions will already set the parent to tx.traceContext if it's zero.

Reply:
I added a test, but I had to expose Spandata.Timestamp for it, I'm not sure if this is the right way.
I also changed apm.StartSpanOptions to your suggestions, but that was not enough because Transaction.StartSpan is not equivalent to Transaction.StartSpanOptions, which I wouldn't know how to fix without breaking the API so added the ended => dropped logic to apm.StartSpanOptions which isn't really great either.

Reply:
> I added a test, but I had to expose Spandata.Timestamp for it, I'm not sure if this is the right way.

The timestamp is meant to be immutable, and exposing it would open up the API to misuse.

The right way to test this would be to get the resulting span event from the recorder transport, and check its timestamp matches what you provided. Take a look at TestTracerStartSpan for an example.

> I also changed apm.StartSpanOptions to your suggestions, but that was not enough because Transaction.StartSpan is not equivalent to Transaction.StartSpanOptions, which I wouldn't know how to fix without breaking the API so added the ended => dropped logic to apm.StartSpanOptions which isn't really great either.

Good point. It's not great repeating the code, but I can find a way to fix it later.

Reply:
I started writing a test and then got carried away and made the other changes too: https://github.com/ChristophPech/apm-agent-go/compare/master...axw:startspanoptions-options

Reply:
This looks great, I'll test it immediately.